### ISSUE TITLE
selector-nested-pattern

### ISSUE THREAD

**Issue by mmacartney10** [stylelint/stylelint#962]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A new rule to only allow pseudo classes to be nested. When I write SASS 

> Which rule, if any, is this issue related to?

e.g. `number-leading-zero`

> What CSS is needed to reproduce this issue?

**Valid**

```css
.item {
	 &:before {
	 	content: '';
	 }
}
```

**Invalid**

```css
.item {
	.child {
		background: #FFF;
	}
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "only-nest-pseudo-classes": true 
  }
}
```

> Which version of stylelint are you using?

`5.0.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Using the PostCss plugin

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

SCSS syntax

**Comment by davidtheclark** [comment]
Yeah, I like this rule and would probably use it. Because we are doing other nesting-specific rules (because there's a CSS spec for it and the pattern's widespread) I'd say we should make it a core rule, right?

**Comment by mmacartney10** [comment]
Yes definitely, the rule would be applicable everywhere so it makes sense for it to be a core rule.

**Comment by mmacartney10** [comment]
There are some cases where someone might want more than just pseudo classes nested. Someone pointed out to me that maybe stateful classes could be allowed. Such as '&.active' or modernizr classes. Could this be included in this rule or is it a case for another rule?

**Comment by davidtheclark** [comment]
Just like we have `selector-class-pattern` we could theoretically have something like `selector-nested-pattern`, where you provide a RegExp to match allowable nested selectors. I suppose that could trump this pseudo-class idea, too, because that rule could cover pseudo-classes as well as arbitrary state classes.

**Comment by mmacartney10** [comment]
Yeah that would be better. Basically for this case I would be only allowing nested items that start or end with an '&'. Is there a reason this cannot be it's own rule where regex doesn't have to be passed through?

**Comment by davidtheclark** [comment]
Your regex would just be very simple!

But the advantage of allowing a regex is that then you *can* do more things: enforce only `&:hover` and `&:focus`, only state class selectors beginning with `&.is-`, etc.

**Comment by mmacartney10** [comment]
That sounds like a good solution, I suppose it better to allow people to have a bit more flexibility in this situation.

**Comment by jeddy3** [comment]
> `selector-nested-pattern`

Sounds great to me. There are lots of different use cases around this and I agree that having a pattern match is the best way to cover them all off.

@mmacartney10 Is this a rule you'd like to contribute? You can look to [`selector-id-pattern`](https://github.com/stylelint/stylelint/blob/master/src/rules/selector-id-pattern/index.js) and [`selector-max-specificity`](https://github.com/stylelint/stylelint/blob/master/src/rules/selector-max-specificity/index.js) for directions. The former will help with the options and documentation, whereas the latter shows how to check each selector in a *selector list*. If you'd like to contribute have a look at [the developer guide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md) and please don't hesitate to give us a holler if you have any questions.

**Comment by mmacartney10** [comment]
@jeddy3 Yeah I would definitely like to contribute this rule, cheers.

**Comment by dryoma** [comment]
Maybe `ignore: ["pseudo-classes", "pseudo-elements"]` option for `max-nesting-depth` would be logical here?

**Comment by jeddy3** [comment]
@dryoma Thanks for the suggestion. I see where you're coming from, but I think combining `max-nesting-depth` with this proposal is would be most prudent course of action because:

- This proposal covers that use case *and more*. Which will allow a user to specify acceptable selector patterns *in one place*, rather than two.
- The [current `max-nesting-depth` ignore option](https://github.com/stylelint/stylelint/tree/master/src/rules/max-nesting-depth#ignore-at-rules-without-declaration-blocks) deals with declaration blocks, which is what rule counts. It's not peering into rules or at-rules, which is what `ignore: ["pseudo-classes", "pseudo-elements"]` would do.

Feel free to champion (i.e. help us build) `selector-nested-pattern` if you'd like to control what selector can and can't be nested in your code base :). [In a previous comment](https://github.com/stylelint/stylelint/issues/962#issuecomment-200409712) I mentioned how to get started.

**Comment by evilebottnawi** [comment]
@jeddy3 i think best solution for this is plugin. Rare usage.

**Comment by jeddy3** [comment]
> Rare usage.

I dunno. I'd use it :)

It also has a clear end-state.

**Comment by woble** [comment]
Out of curiosity, would this new proposed rule propagate to `selector-class-pattern` when `resolveNestedSelectors` is set to `true`?

**Comment by jeddy3** [comment]
> Out of curiosity, would this new proposed rule propagate to selector-class-pattern when resolveNestedSelectors is set to true?

Nope. `resolveNestedSelectors` is a secondary option set on per-rule basis. It doesn't make sense for this rule to have that option.

You'd use this rule to enforce, for example, that all nested selectors (whether they are classes, id or pseudos etc) need to be preceded by a `&`. You'd then use the likes of `selector-id-pattern` and `selector-id-pattern` (with their `resolveNestedSelectors` secondary options set to true) to specify what those classes and id selectors should look it.
